modules/wechat_qrcode/src/wechat_qrcode.cpp

```diff
@@ -0,0 +1,205 @@
+    bool use_nn_detector_, use_nn_sr_;
+};
+
+WeChatQRCode::WeChatQRCode(const String& detector_prototxt_path,
+                           const String& detector_caffe_model_path,
+                           const String& super_resolution_prototxt_path,
+                           const String& super_resolution_caffe_model_path) {
+    p = makePtr<WeChatQRCode::Impl>();
+    if (!detector_caffe_model_path.empty() && !detector_prototxt_path.empty()) {
+        // initialize detector model (caffe)
+        p->use_nn_detector_ = true;
+        CV_Assert(utils::fs::exists(detector_prototxt_path));
+        CV_Assert(utils::fs::exists(detector_caffe_model_path));
+        p->detector_ = make_shared<SSDDetector>();
+        auto ret = p->detector_->init(detector_prototxt_path, detector_caffe_model_path);
+        CV_Assert(ret == 0);
+    } else {
+        p->use_nn_detector_ = false;
+        p->detector_ = NULL;
+    }
+    // initialize super_resolution_model
+    // it could also support non model weights by cubic resizing
+    // so, we initialize it first.
+    p->super_resolution_model_ = make_shared<SuperScale>();
+    if (!super_resolution_prototxt_path.empty() && !super_resolution_caffe_model_path.empty()) {
+        p->use_nn_sr_ = true;
+        // initialize dnn model (caffe format)
+        CV_Assert(utils::fs::exists(super_resolution_prototxt_path));
+        CV_Assert(utils::fs::exists(super_resolution_caffe_model_path));
+        auto ret = p->super_resolution_model_->init(super_resolution_prototxt_path,
+                                                    super_resolution_caffe_model_path);
+        CV_Assert(ret == 0);
+    } else {
+        p->use_nn_sr_ = false;
+    }
+}
+
+vector<string> WeChatQRCode::detectAndDecode(InputArray img, OutputArrayOfArrays points) {
+    CV_Assert(!img.empty());
+    CV_CheckDepthEQ(img.depth(), CV_8U, "");
+
+    if (img.cols() <= 20 || img.rows() <= 20) {
+        return vector<string>();  // image data is not enough for providing reliable results
+    }
+    Mat input_img;
+    int incn = img.channels();
+    CV_Check(incn, incn == 1 || incn == 3 || incn == 4, "");
+    if (incn == 3 || incn == 4) {
+        cvtColor(img, input_img, COLOR_BGR2GRAY);
+    } else {
+        input_img = img.getMat();
+    }
+    auto candidate_points = p->detect(input_img);
+    auto res_points = vector<Mat>();
+    auto ret = p->decode(input_img, candidate_points, res_points);
+    // opencv type convert
+    vector<Mat> tmp_points;
+    if (points.needed()) {
+        for (size_t i = 0; i < res_points.size(); i++) {
+            Mat tmp_point;
+            tmp_points.push_back(tmp_point);
+            res_points[i].convertTo(((OutputArray)tmp_points[i]),
+                                    ((OutputArray)tmp_points[i]).fixedType()
+                                        ? ((OutputArray)tmp_points[i]).type()
+                                        : CV_32FC2);
```

Comment:
`.fixedType()` is used with external arguments only (mentioned file from `objdetect` follows this rule).
No sense to use it with "local" variables.

---

> good ideas

Reuse conventions from external API of objdetect/QRCodeDetector

Reply:
OK. I will remove it.

Reply:
@alalek  Could we reopen this PR?